packages/eslint-plugin/src/rules/no-circular-import.ts

```diff
@@ -0,0 +1,181 @@
+import * as ts from 'typescript';
+
+import { createRule, getParserServices } from '../util';
+
+type Options = [];
+type MessageIds = 'noCircularImport';
+
+interface Edge {
+  filename: string;
+  specifier: string;
+}
+
+class Graph {
+  private graph = new Map<
+    string,
+    {
+      filename: string;
+      specifier: string;
+    }[]
+  >();
+
+  addEdge(start: string, edge: Edge): void {
+    if (this.graph.has(start)) {
+      this.graph.get(start)?.push(edge);
+    } else {
+      this.graph.set(start, [edge]);
+    }
+  }
+
+  hasEdge(name: string): boolean {
+    return this.graph.has(name);
+  }
+
+  getEdges(name: string): Edge[] {
+    return this.graph.get(name) ?? [];
+  }
+}
+//  imports “a.ts” and is imported from “b.ts”, resulting in a circular reference.
+export default createRule<Options, MessageIds>({
+  name: 'no-circular-import',
+  meta: {
+    docs: {
+      description:
+        'Disallow the use of import module that result in circular imports',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      noCircularImport: 'Circular import dcetected via {{paths}}.',
+    },
+    schema: [],
+    type: 'suggestion',
+  },
+  defaultOptions: [],
+  create(context) {
+    const services = getParserServices(context);
+    const graph = new Graph();
```

Comment:
[Performance] This is what we've been afraid of 😬. Creating a new graph on each file becomes expensive in larger projects. From https://github.com/typescript-eslint/typescript-eslint/issues/224#issuecomment-1116455239:

> We can rely on the fact that TS has already analysed every file and knows the dependency graph of the project.
> So instead of manually building it - we can just ask TS for the information.

Having to manually recreate the graph on demand is not what I was hoping we'd have to do here. I don't see any TypeScript APIs that can give the full dependency graph on-demand.

I tried out this rule locally with a basic config and it increased the lint time from ~35-36 seconds to ~39-40 seconds on my M1 Mac. So, not _immediately terrible_... but noticeable, and will get exponentially worse in larger projects.

<details>
<summary>Testing local <code>eslint.config.mjs</code></summary>

```js
import tseslint from 'typescript-eslint';
import deprecationPlugin from 'eslint-plugin-deprecation';
import tseslintInternalPlugin from '@typescript-eslint/eslint-plugin-internal';
import eslintCommentsPlugin from 'eslint-plugin-eslint-comments';
import eslintPluginPlugin from 'eslint-plugin-eslint-plugin';
import importPlugin from 'eslint-plugin-import';
import jestPlugin from 'eslint-plugin-jest';
import jsdocPlugin from 'eslint-plugin-jsdoc';
import jsxA11yPlugin from 'eslint-plugin-jsx-a11y';
import reactPlugin from 'eslint-plugin-react';
import reactHooksPlugin from 'eslint-plugin-react-hooks';
import simpleImportSortPlugin from 'eslint-plugin-simple-import-sort';
import unicornPlugin from 'eslint-plugin-unicorn';

export default tseslint.config(
  {
    ignores: [
      '**/__snapshots__/**',
      '**/.docusaurus/**',
      '**/.nx/**',
      '**/build/**',
      '**/coverage/**',
      '**/dist/**',
      '**/fixtures/**',
      '**/jest.config.js',
      '**/node_modules/**',
      'packages/rule-tester/tests/eslint-base',
      'packages/types/src/generated/**/*.ts',
      'packages/website/src/vendor',
    ],
  },
  ...tseslint.configs.recommended,
  {
    languageOptions: {
      parserOptions: {
        EXPERIMENTAL_useProjectService: {
          allowDefaultProjectForFiles: ['./*.js', './*/*.js', './*/*/*.js'],
          defaultProject: 'tsconfig.json',
        },
      },
    },
    plugins: {
      ['@typescript-eslint']: tseslint.plugin,
      ['@typescript-eslint/internal']: tseslintInternalPlugin,
      ['deprecation']: deprecationPlugin,
      ['eslint-comments']: eslintCommentsPlugin,
      ['eslint-plugin']: eslintPluginPlugin,
      ['import']: importPlugin,
      ['jest']: jestPlugin,
      ['jsdoc']: jsdocPlugin,
      ['jsx-a11y']: jsxA11yPlugin,
      ['react-hooks']: reactHooksPlugin,
      ['react']: reactPlugin,
      ['simple-import-sort']: simpleImportSortPlugin,
      ['unicorn']: unicornPlugin,
    },
    rules: {
      '@typescript-eslint/ban-ts-comment': 'off',
      '@typescript-eslint/ban-types': 'off',
      '@typescript-eslint/no-array-constructor': 'off',
      '@typescript-eslint/no-duplicate-enum-values': 'off',
      '@typescript-eslint/no-explicit-any': 'off',
      '@typescript-eslint/no-extra-non-null-assertion': 'off',
      '@typescript-eslint/no-loss-of-precision': 'off',
      '@typescript-eslint/no-misused-new': 'off',
      '@typescript-eslint/no-namespace': 'off',
      '@typescript-eslint/no-non-null-asserted-optional-chain': 'off',
      '@typescript-eslint/no-this-alias': 'off',
      '@typescript-eslint/no-unnecessary-type-constraint': 'off',
      '@typescript-eslint/no-unsafe-declaration-merging': 'off',
      '@typescript-eslint/no-unused-vars': 'off',
      '@typescript-eslint/no-var-requires': 'off',
      '@typescript-eslint/prefer-as-const': 'off',
      '@typescript-eslint/triple-slash-reference': 'off',
      'no-async-promise-executor': 'off',
      'no-case-declarations': 'off',
      'no-constant-condition': 'off',
      'no-empty': 'off',
      'no-fallthrough': 'off',
      'no-prototype-builtins': 'off',
      'no-undef': 'off',
      'no-unsafe-finally': 'off',

      '@typescript-eslint/no-circular-import': 'error',
    },
  },
);
```
</details>

I question whether this really makes sense to include in our plugin with this _"create a graph for each file"_ architecture.  Similar to how #371 was closed in favor of external tooling like [Knip](https://knip.dev), maybe this just doesn't belong in a lint rule?